core/src/apps/webauthn/__init__.py

```diff
@@ -749,3 +1230,413 @@ def msg_error(cid: int, code: int) -> Cmd:
+    else:
+        # Ask user to confirm registration.
+        state_set = dialog_mgr.set_state(
+            Fido2ConfirmMakeCredential(
+                req.cid,
+                dialog_mgr.iface,
+                client_data_hash,
+                cred,
+                resident_key,
+                user_verification,
+            )
+        )
+
+    if not state_set:
+        return cmd_error(req.cid, _ERR_CHANNEL_BUSY)
+
+    return None
+
+
+def cbor_make_credential_sign(
+    client_data_hash: bytes, cred: Fido2Credential, user_verification: bool
+) -> bytes:
+    privkey = cred.private_key()
+    pubkey = nist256p1.publickey(privkey, False)
+
+    flags = _AUTH_FLAG_UP | _AUTH_FLAG_AT
+    if user_verification:
+        flags |= _AUTH_FLAG_UV
+
+    # Encode the authenticator data (Credential ID, its public key and extensions).
+    credential_pub_key = cbor.encode(
+        {
+            _COSE_ALG_KEY: _COSE_ALG_ES256,
+            _COSE_KEY_TYPE_KEY: _COSE_KEY_TYPE_EC2,
+            _COSE_CURVE_KEY: _COSE_CURVE_P256,
+            _COSE_X_COORD_KEY: pubkey[1:33],
+            _COSE_Y_COORD_KEY: pubkey[33:],
+        }
+    )
+
+    att_cred_data = (
+        _AAGUID + len(cred.id).to_bytes(2, "big") + cred.id + credential_pub_key
+    )
+
+    extensions = b""
+    if cred.hmac_secret:
+        extensions = cbor.encode({"hmac-secret": True})
+        flags |= _AUTH_FLAG_ED
+
+    authenticator_data = (
+        cred.rp_id_hash
+        + bytes([flags])
+        + b"\x00\x00\x00\x00"
+        + att_cred_data
+        + extensions
+    )
+
+    # Compute self-attestation signature of the authenticator data.
+    dig = hashlib.sha256()
+    dig.update(authenticator_data)
+    dig.update(client_data_hash)
+    sig = nist256p1.sign(privkey, dig.digest(), False)
+    sig = der.encode_seq((sig[1:33], sig[33:]))
+
+    # Encode the authenticatorMakeCredential response data.
+    return cbor.encode(
+        {
+            _MAKECRED_RESP_FMT: "packed",
+            _MAKECRED_RESP_AUTH_DATA: authenticator_data,
+            _MAKECRED_RESP_ATT_STMT: {"alg": _COSE_ALG_ES256, "sig": sig},
+        }
+    )
+
+
+def cbor_get_assertion(req: Cmd, dialog_mgr: DialogManager) -> Optional[Cmd]:
+    if not storage.is_initialized():
+        if __debug__:
+            log.warning(__name__, "not initialized")
+        return cbor_error(req.cid, _ERR_OPERATION_DENIED)
+
+    try:
+        param = cbor.decode(req.data[1:])
+        rp_id = param[_GETASSERT_CMD_RP_ID]
+        rp_id_hash = hashlib.sha256(rp_id).digest()
+
+        allow_list = param.get(_GETASSERT_CMD_ALLOW_LIST, [])
+        if allow_list:
+            # Get all credentials from the allow list that belong to this authenticator.
+            cred_list = []
+            for credential_descriptor in allow_list:
+                if credential_descriptor["type"] != "public-key":
+                    continue
+                cred = Credential.from_bytes(credential_descriptor["id"], rp_id_hash)
```

Comment:
Neither CTAP2 nor Webauthn says nothing about processing U2F credentials here.

Reply:
I suggest warning user if he or she is going to use U2F credentials.

Reply:
Can you describe the attack scenario that this would mitigate against?

Reply:
I suggested it for the following reasons:

1. A user doesn't expect that the token use U2F credential for FIDO2 authentication (or FIDO2 credential for U2F authentication).
2. It's a feature that doesn't comply with FIDO2, moreover from cryptography point of view it makes sense to separate space of FIDO2 credentials and U2F credentials.
3. It's a trade-off between allowing and not allowing mixing U2F and FIDO2 credentials.

I can imagine various forms of the warning. For example, a text indicating whether the credential a user is going to use is FIDO2 or U2F.

Reply:
I don't think I can judge point 1.

Regarding point 2: In general I of course agree that credentials generated using different protocols should be domain separated. I don't have a problem if credentials generated using an older version of a protocol are used in a newer version of the same protocol if the protocol takes that into account, such as described in section 7 of the CTAP2 spec ([Interoperating with CTAP1/U2F authenticators](https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#u2f-interoperability)). I am generally not happy about the opposite scenario, which could be called a "protocol downgrade", but the current browser and/or relying party implementations require that we support it here and the fact of the matter is that we are not aware of any attacks that can be mounted in case of this downgrade, since (currently) the difference between the two protocols is neither in the cryptography we support nor in the security-critical behavior, but rather in FIDO2 being more feature-rich. I don't recall reading that mixing U2F and FIDO2 credentials is non-compliant with the specs. @onvej can you reference that bit? I am not sure how that would go with CTAP2 Section 7.

Regarding the warning: I don't like the idea of showing the user a warning, which to our best knowledge does not serve any meaningful purpose. I wouldn't want users contacting support and us telling them "oh yeah, you can ignore that", because that sets a very bad example. I am OK with showing the credential type, if we can fit it on the screen.

@prusnak @hiviah lets give this some thought and discuss tomorrow.